### PR TITLE
Fix #1543: Introduce the ES6 ...spread operator in the IR.

### LINKED ISSUE #1543: ES6 ...spread operator
ECMAScript 6 introduces a [spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_operator), which is basically the equivalent of Scala's `: _*` to pass an iterable as varargs to a JS function (or within an array).

The very nice thing about this operator is that removes the need the `apply` method of JS functions (unless you want to bind `this`, like when calling `js.ThisFunction`s.). Since it can also be used when invoking *constructors*, we can also push the logic of calling [newJSObjectWithVarArgs](https://github.com/scala-js/scala-js/blob/v0.6.1/library/src/main/scala/scala/scalajs/runtime/package.scala#L40-L53) from the compiler to desugar (in ES 5) and to `...` (in ES 6).

That would be a non-forward compatible change in the IR.

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0, but NOT MERGE. This must not go into 0.6.2.

File: library/src/main/scala/scala/scalajs/runtime/package.scala [hunk 1]
```diff
@@ -37,20 +37,16 @@ package object runtime {
...
-        result
-    }
-  }
+  /** Instantiates a JS object with variadic arguments to the constructor.
+   *
+   *  This method is kept for backward binary compatibility with 0.6.{0,1},
+   *  whose codegen would generate calls to this method when instantiating
+   *  a JS object with variadic arguments.
+   */
+  @deprecated("Was generated by the codegen < 0.6.2, now replaced by the spread operator.", "0.6.2")
```
Comment (gzm0) [inline library/src/main/scala/scala/scalajs/runtime/package.scala]: The comment is not clear here. You mean this was needed by the codegen.

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala [hunk 2]
```diff
@@ -3277,16 +3225,18 @@ abstract class GenJSCode extends plugins.PluginComponent
...
-     *  Returns an only tree which is a JS array of the arguments. In most
-     *  cases, it will be a js.JSArrayConstr with the expanded arguments. It will
-     *  not if a Seq is passed to a varargs argument with the syntax seq: _*.
+     *  Repeated arguments that cannot be expanded at compile time (i.e., if a
+     *  Seq is passed to a varargs parameter with the syntax `seq: _*`) will be
+     *  wrapped in a [[js.JSSpread]] node to be expanded at runtime.
      */
     private def genPrimitiveJSArgs(sym: Symbol, args: List[Tree])(
-        implicit pos: Position): js.Tree = {
+        implicit pos: Position): List[js.Tree] = {
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: In this method you are assuming that none of the non-last individual parameters result in a `JSSpread`. If this is not the case, we will produce bad IR. Maybe we should explicitly guard against that.

Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: Oooohhhhhh.... spread is allowed anywhere, right? Sweeeet :)

File: tools/scalajsenv.js [hunk 3]
```diff
@@ -341,6 +341,19 @@ ScalaJS.isInfinite = function(instance) {
...
+  var instance = ScalaJS.g["Object"]["create"](ctor.prototype);
+  var result = ctor["apply"](instance, args);
+  switch (typeof result) {
+    case "string": case "number": case "boolean": case "undefined": case "symbol":
+      return instance;
+    default:
+      return result === null ? instance : result;
+  }
+};
+
```
Comment (gzm0) [inline tools/scalajsenv.js]: Aaaaand it's in JavaScript again :)

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala (hunk 2 above)
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: Yes. Rest parameters are only allowed at the end. But spread is allowed anywhere, and can be used more than once.

File: tools/scalajsenv.js (hunk 3 above)
Comment (sjrd) [inline tools/scalajsenv.js]: Yep ^^

**Comment by gzm0** [comment]
That's all

**Comment by gzm0** [comment]
This is such a beautiful PR.

**Comment by gzm0** [comment]
(except for `spreadToArgArray` I've always hated this code... I spent 30mins thinking of something better... in vain :()

**Comment by sjrd** [comment]
My sentiment exactly.

**Comment by sjrd** [comment]
Updated.

**Comment by gzm0** [comment]
LGTM

**Comment by gzm0** [comment]
test this please

**Comment by gzm0** [comment]
Something strange is going on with Mima... it seems to try and resolve 0.6.3-SNAPSHOT...

**Comment by sjrd** [comment]
Updated. The binary version for MiMa was computed from the binary version of the current version, instead of using the binary version of the previous version to be compared to. But then resolving the irJS and toolsJS artifacts failed.

**Comment by gzm0** [comment]
The fix to the build file should have gone into a separate commit :-/

**Comment by sjrd** [comment]
It was linked to the change of binary version, which was linked to the change of the IR.

**Comment by gzm0** [comment]
> It was linked to the change of binary version

No, not really. The binary version of the previous artifact is always the binary version of the previous artifact. It was a logically flawed piece of code, whose flaw manifested due to the binary version change.

**Comment by sjrd** [comment]
Hum. Yes you're right.

**Pull request merged**